### PR TITLE
[Parser] Don't modify the current token kind when cutting off parsing

### PR DESCRIPTION
**Pull request by ahoppen**
Instead, let the lexer know that it should interpret the next token to be lexed as an EOF token. This makes parsing cutoff a lot more stable because the parser is not expecting the current token kind to change in a lot of places.

Technically, this causes two more tokens to be parsed than in the previous implementation (namely `Parser.Tok` and `Lexer.NextToken`) but since we bail out at a fairly low nesting level, this shouldn't make a big difference.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: validation-test/compiler_crashers_2_fixed/parser-cutoff.swift [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+if a > 1 {
+if a > 1 {
+if a > 1 {
+if a > 1 {
+if a > 1 {
+if a > 1 {
+if a > 1 {
+if a > 1 {
+if a > 1 {
+
```
Comment (bnbarham) [inline validation-test/compiler_crashers_2_fixed/parser-cutoff.swift:50]: I assume this was from me, obviously doesn't matter but could remove if you wanted it to be a little neater :P

**Review by bnbarham — APPROVED** [review]
Nice!

File: validation-test/compiler_crashers_2_fixed/parser-cutoff.swift (hunk 1 above)
Comment (ahoppen) [reply validation-test/compiler_crashers_2_fixed/parser-cutoff.swift:50]: Whatever. The file hit the issue before and now doesn’t hit it any longer so I don’t want to change it anymore.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
I refactored the PR a bit, reverting a fair bit of https://github.com/apple/swift/pull/19631 because the checks for `StructureMarkerRAII::isFailed` are no longer necessary with the new cut-off model.

@bnbarham Could you take another look at the changes?

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: validation-test/compiler_crashers_2_fixed/parser-cutoff.swift [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+// RUN: not %swift-frontend -typecheck %s
```
Comment (bnbarham) [inline validation-test/compiler_crashers_2_fixed/parser-cutoff.swift:791]: The crash here was the assertion failure right? Would it be worth adding `// REQUIRES: asserts` or a comment to that effect in here? The test does work either way so maybe doesn't matter, but IMO it'd be nice to differentiate it from the existing `structure_overflow*.swift` ones.

**Review by bnbarham — APPROVED** [review]
Nice, much cleaner!

File: validation-test/compiler_crashers_2_fixed/parser-cutoff.swift (hunk 2 above)
Comment (ahoppen) [reply validation-test/compiler_crashers_2_fixed/parser-cutoff.swift:792]: Added a comment.

File: lib/Parse/Parser.cpp [hunk 3]
```diff
@@ -405,6 +405,9 @@ namespace {
 /// underlying corrected token stream.
 class TokenRecorder: public ConsumeTokenReceiver {
   ASTContext &Ctx;
+  /// The lexer that is being used to lex the source file. Used to query whether
+  /// lexing has been cut off.
+  Lexer *BaseLexer;
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply lib/Parse/Parser.cpp]: Good idea. Changed it to a reference.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Pull request merged**